flow/aggregates.py

```diff
@@ -0,0 +1,266 @@
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from collections.abc import Iterable
+from hashlib import md5
+from hashlib import sha1
+from itertools import groupby
+from itertools import zip_longest
+
+
+class Aggregate:
+    """Decorator for operation functions that are to be aggregated.
+    By default, if the aggregator parameter is not passed,
+    an aggregate of all jobs will be created.
+
+    .. code-block:: python
+
+        example_aggregate = Aggregate()
+        @example_aggregate
+        @FlowProject.operation
+        def foo(*jobs):
+            print(len(jobs))
+
+    :param aggregator:
+        Information describing how to aggregate jobs. Is a callable that
+        takes in a list of jobs and can return or yield subsets of jobs as
+        an iterable. The default behavior is creating a single aggregate
+        of all jobs
+    :type aggregator:
+        callable
+    :param sort_by:
+        Before aggregating, sort the jobs by a given statepoint parameter.
+        The default behavior is no sorting.
+    :type sort_by:
+        str or NoneType
+    :param reverse_order:
+        States if the jobs are to be sorted in reverse order.
+        The default value is False.
+    :type reverse_order:
+        bool
+    :param select:
+        Condition for filtering individual jobs. This is passed as the
+        callable argument to `filter`.
+        The default behavior is no filtering.
+    :type select:
+        callable or NoneType
+    """
+
+    def __init__(self, aggregator=None, sort_by=None, reverse_order=False, select=None):
+        if aggregator is None:
+            def aggregator(jobs):
+                return [jobs]
+
+        if not callable(aggregator):
+            raise TypeError(f"Expected callable for aggregator, got {type(aggregator)}")
+
+        if sort_by is not None and not isinstance(sort_by, str):
+            raise TypeError(f"Expected string sort_by parameter, got {type(sort_by)}")
+
+        if select is not None and not callable(select):
+            raise TypeError(f"Expected callable for select, got {type(select)}")
+
+        if sort_by is None and select is None and not reverse_order:
+            self._is_aggregate = getattr(aggregator, '_num', 0) != 1
+        else:
+            self._is_aggregate = True
+
+        self._aggregator = aggregator
+        self._sort_by = sort_by
+        self._reverse_order = reverse_order
+        self._select = select
+
+    @classmethod
+    def groupsof(cls, num=1, sort_by=None, reverse_order=False, select=None):
+        # copied from: https://docs.python.org/3/library/itertools.html#itertools.zip_longest
+        try:
+            num = int(num)
+            if num <= 0:
+                raise ValueError('The num parameter should have a value greater than 0')
+        except TypeError:
+            raise TypeError('The num parameter should be an integer')
```

Comment:
@bdice another instance of type checking up front that is probably justified due to the delayed execution of aggregator.

@kidrahahjo the sequence here is a little confusing though. The `TypeError` you're catching would be thrown by the first line in the `try` block, so there's no need for the rest to be there. I would move the `if num<=0` check out of the try block. Furthermore, rather than trying to convert to int like this, if we're going to type check I would do it rigorously. Currently, passing `num=3.5` will result in the code silently using `num=3`.
```suggestion
        if num != int(num):
            raise TypeError('The num parameter should be an integer')
        num = int(num)
        if num <= 0:
            raise ValueError('The num parameter should have a value greater than 0')
```